doc/memos/rdm-draft-alamos-ieee802154-radio-hal.md

```diff
@@ -0,0 +1,908 @@
+     * dynamic framebuffer protection or disabled RX). Thus, the packet
+     * will not be overwritten before calling the @ref ieee802154_radio_read
+     * function. However, @ref ieee802154_radio_read MUST be called in
+     * order to receive new packets. If there is no interest in the
+     * packet, the function can be called with a NULL buffer to drop
+     * the packet.
+     */
+    IEEE802154_RADIO_RX_DONE,
+
+    /**
+     * @brief the transceiver finished sending a packet or the
+     *        retransmission procedure.
+     *
+     * If the radio supports frame retransmissions the
+     * @ref ieee802154_radio_get_tx_status MAY be called to retrieve useful
+     * information (number of retries, frame pending bit, etc). The
+     * transceiver is in @ref IEEE802154_TRX_STATE_TX_ON state when this function
+     * is called.
+     */
+    IEEE802154_RADIO_TX_DONE,
+} ieee802154_trx_status_t;
+
+/**
+ * @brief Prototype of the IEEE802.15.4 device event callback
+ *
+ * @param[in] dev IEEE802.15.4 device descriptor
+ * @param[in] status the event status
+ */
+typedef void (*ieee802154_cb_t)(ieee802154_dev_t *dev,
+                                ieee802154_trx_status_t status);
+```
+
+Other MAC specific events such as TX done with frame pending, CSMA-CA medium
+busy or exceeded number of retransmissions are not explicitly reported because
+they can be extracted after the TX done event using the Radio HAL API.
+
+The Radio HAL is designed to be able to call the Radio Ops interface from the
+event notification callback.
+
+All Radio HAL implementation MUST implement at least the `RX_START`, `RX_DONE`
+and `TX_DONE` events. All these event might be disabled via compile time
+configurations if the upper layer doesn't require them.
+
+Other events such as `ACK_TIMEOUT` in radios that support ACK timers
+and `CSMA_BACKOFF_TIMEOUT` in radios that support CSMA-CA backoff timers are
+out of the scope of this document. However, these might be added and implemented
+at any time if required by upper layers.
+
+## 6.3 Radio Capabilities
+The following list defines the basic capabilities available in common
+IEEE802.15.4 compliant radios. Other capabilities such as support for CSMA-CA
+backoff and ACK timers are not included in the list because they are out of
+the scope of this document.
+
+```c
+typedef enum {
+    /**
+     * @brief the device supports hardware address filter
+     */
+    IEEE802154_CAP_HW_ADDR_FILTER,
+    /**
+     * @brief the device support frame retransmissions with CSMA-CA
+     */
+    IEEE802154_CAP_FRAME_RETRIES,
+    /**
+     * @brief the device support ACK timeout interrupt
+     */
+    IEEE802154_CAP_ACK_TIMEOUT,
+    /**
+     * @brief the device supports Auto ACK
+     */
+    IEEE802154_CAP_AUTO_ACK,
+    /**
+     * @brief the device performs CCA when sending
+     */
+    IEEE802154_CAP_24_GHZ,
+    /**
+     * @brief the device support the IEEE802.15.4 Sub GHz band
+     */
+    IEEE802154_CAP_SUB_GHZ,
```

Comment:
`IEEE802154_PHY_MODES` could be another capability.

Reply:
yes, this can be used to report available PHY modes too

Reply:
actually, this list can/should be expanded. There are other capabilities that are not listed here but exist anyway (e.g LBT, CSMA_BACKOFF timers, etc)

Reply:
The list can and should be extended as explained [here](https://github.com/RIOT-OS/RIOT/pull/13943/files/03dc1f9bf97016c10680201609fe815ef4cd9c21#diff-148efb534ddd3b28a50a11d894261922R887). Personally, I hope that it won't get extended as much as `netopt`.

Reply:
maybe we could try to limit the scope here. Usually, capabilities are mapped to hardware accelerations, features and certain conditions (e.g if the radio calculates the checksum, that's a CAP).

In practice, I've seen that this list doesn't explode (see [here](https://github.com/zephyrproject-rtos/zephyr/blob/a9eda4f622f838f066e992c2a2d4dff4ba18ff32/include/net/ieee802154_radio.h#L29) or [here](https://elixir.bootlin.com/linux/latest/source/include/net/mac802154.h#L118))